docs/architecture/adr-061-core-module-api.md

```diff
@@ -0,0 +1,490 @@
+defined here. In the case of gogo proto and amino interfaces, the registration of these generally should happen as early
+as possible during initialization and in [ADR 057: App Wiring](./adr-057-app-wiring-1.md), protobuf type registration  
+happens before dependency injection (although this could alternatively be done dedicated DI providers).
+
+Commands should likely be handled at a different level of the framework as they are purely a client concern. Currently,
+we use the cobra framework, but there have been discussions of potentially using other frameworks, automatically
+generating CLI commands, and even letting libraries outside the SDK totally manage this.
+
+gRPC gateway registration should probably be handled by the runtime module, but the core API shouldn't depend on gRPC
+gateway types as 1) we are already using an older version and 2) it's possible the framework can do this registration
+automatically in the future. So for now, the runtime module should probably provide some sort of specific type for doing
+this registration ex:
+```go
+type GrpcGatewayInfo struct {
+	Handlers []GrpcGatewayHandler
+}
+
+type GrpcGatewayHandler func(ctx context.Context, mux *runtime.ServeMux, client QueryClient) error
+```
+
+which modules can return in a provider:
+```go
+func ProvideGrpcGateway() GrpcGatewayInfo {
+	return GrpcGatewayinfo {
+        Handlers: []Handler {types.RegisterQueryHandlerClient}
+	}
+}
+```
+
+Crisis module invariants and simulations are subject to potential redesign and should be managed with types
+defined in the crisis and simulation modules respectively.
+
+#### Example Usage
+
+Here is an example of setting up a hypothetical `foo` v2 module which uses the [ORM](./adr-055-orm.md) for its state
+management and genesis.
+
+```go
+func ProvideApp(config *foomodulev2.Module, evtSvc event.EventService, db orm.ModuleDB) (Keeper, *Handler){
+	k := &Keeper{db: db, evtSvc: evtSvc}
+	h := &Handler{}
+	foov1.RegisterMsgServer(h, k)
+    foov1.RegisterQueryServer(h, k)
+	h.RegisterBeginBlocker(k.BeginBlock)
+    db.RegisterGenesis(h)
+    h.RegisterUpgradeHandler("foo.module.v1.Module", k.MigrateFromV1)
+	return k, h
+}
+```
+
+### Runtime Compatibility Version
+
+The `core` module will define a static integer var, `cosmossdk.io/core/RuntimeCompatibilityVersion`, which is
+a minor version indicator of the core module that is accessible at runtime. Correct runtime module implementations
+should check this compatibility version and return an error if the current `RuntimeCompatibilityVersion` is higher
+than the version of the core API that this runtime version can support. When new features are adding to the `core`
+module API that runtime modules are required to support, this version should be incremented.
+
+## Consequences
+
+### Backwards Compatibility
+
+Early versions of runtime modules should aim to support as much as possible modules built with the existing
```

Comment:
The wording here is runtime _modules_ but in practice the SDK team will implement just one, at least initially, is that right?

I'm having a hard time visualizing how runtime would straddle the current `AppModule / sdk.Context` paradigm and   the proposed composition through `Handler`.

Reply:
Well, I think we would implement one for the existing tendermint version, and another one once ABCI++ comes around possibly deprecating `AppModule` altogether. The legacy version might have a fairly long life span of patch updates.

As for the second part, you would just have some sort of module manager which can deal with both `AppModule` and `Handler` appropriately.